lib/doh.c

```diff
@@ -894,8 +1014,239 @@ UNITTEST void de_cleanup(struct dohentry *d)
+   * string values, so we might see a comma within the wire format
+   * version of a string, in which case we'll precede that by a
+   * backslash - same goes for a backslash character
+   */
+  int remaining = (int) len;
+  char *oval, *op;
+  unsigned char *cp = rrval;
+
+  if(!alpns)
+    return CURLE_OUT_OF_MEMORY;
+  /* allow space to quote possibly every single character :-) */
+  oval = calloc(1, 2 * len);
+  if(!oval)
+    return CURLE_OUT_OF_MEMORY;
+  op = oval;
+  while(remaining > 0) {
+    size_t tlen = (size_t) *cp++;
+    remaining--;
+    if(tlen > (size_t)remaining || tlen > (size_t)((2 * len) - (op - oval)))
+      goto err;
+    memcpy(op, cp, tlen);
+    op += tlen;
+    if((size_t)(op - oval) >= ((size_t)2 * len))
+      goto err;
+    *op++ = COMMA_CHAR;
+    cp += tlen;
+    remaining -= (int)tlen;
+  }
+  oval[strlen(oval) - 1] = '\0';
+  /* TODO: handle escaping */
+
+  *alpns = oval;
+  return CURLE_OK;
+err:
+  if(oval)
+    free(oval);
+  return CURLE_BAD_CONTENT_ENCODING;
+}
+
+static CURLcode Curl_doh_decode_httpsrr(unsigned char *rrval, size_t len,
+                                        struct Curl_https_rrinfo **hrr)
+{
+  size_t remaining = len;
+  unsigned char *cp = rrval;
+  uint16_t pcode = 0, plen = 0;
+  struct Curl_https_rrinfo *lhrr = NULL;
+  char *dnsname = NULL;
+
+  lhrr = calloc(1, sizeof(struct Curl_https_rrinfo));
+  if(!lhrr)
+    return CURLE_OUT_OF_MEMORY;
+  lhrr->val = calloc(1, len);
+  if(!lhrr->val)
+    goto err;
+  lhrr->len = len;
+  memcpy(lhrr->val, rrval, len);
+  if(remaining <= 2)
+    goto err;
+  lhrr->priority = (uint16_t)((cp[0] << 8) + cp[1]);
+  cp += 2;
+  remaining -= (uint16_t)2;
+  if(local_decode_rdata_name(&cp, &remaining, &dnsname) != CURLE_OK)
+    goto err;
+  lhrr->target = dnsname;
+  while(remaining >= 4) {
+    pcode = (uint16_t)((*cp << 8) + (*(cp + 1)));
+    cp += 2;
+    plen = (uint16_t)((*cp << 8) + (*(cp + 1)));
+    cp += 2;
+    remaining -= 4;
+    if(pcode == HTTPS_RR_CODE_ALPN) {
+      if(local_decode_rdata_alpn(cp, plen, &lhrr->alpns) != CURLE_OK)
+        goto err;
+    }
+    if(pcode == HTTPS_RR_CODE_NO_DEF_ALPN)
+      lhrr->no_def_alpn = TRUE;
+    if(pcode == HTTPS_RR_CODE_IPV4) {
+      lhrr->ipv4hints = calloc(1, plen);
+      if(!lhrr->ipv4hints)
+        goto err;
+      lhrr->ipv4hints_len = (size_t)plen;
+      memcpy(lhrr->ipv4hints, cp, plen);
+    }
+    if(pcode == HTTPS_RR_CODE_ECH) {
+      lhrr->echconfiglist = calloc(1, plen);
+      if(!lhrr->echconfiglist)
+        goto err;
+      lhrr->echconfiglist_len = (size_t)plen;
+      memcpy(lhrr->echconfiglist, cp, plen);
+    }
+    if(pcode == HTTPS_RR_CODE_IPV6) {
+      lhrr->ipv6hints = calloc(1, plen);
+      if(!lhrr->ipv6hints)
+        goto err;
+      lhrr->ipv6hints_len = (size_t)plen;
+      memcpy(lhrr->ipv6hints, cp, plen);
+    }
+    if(plen > 0 && plen <= remaining) {
+      cp += plen;
+      remaining -= plen;
+    }
+  }
```

Comment:
If `remaining` is non-zero at this point, something is likely wrong in the parser, right? Maybe you can just add a `DEBUGASSERT(!remaining);` there to trigger a debug build to notice this for further research?

Reply:
done